### DISCUSSION TITLE
The next step for Code Review GPT

### DISCUSSION THREAD

**Discussion by mattzcarey** [mattzcarey/shippie#338] [Announcements]
Hey everyone,

Thanks for all your interest and support of this package. Lots of changes have been happening in my life in the last 6 months hence the stall in progress. I wanted to share some updates on Code Review GPT and my thoughts about code review tools more widely.

Looking at the current state of market of these AI-powered review tools as of July 2024, Code Review GPT is up against increasingly sophisticated alternatives such as CodeRabbit and Codium. These companies are raising huge amounts of money to automate away the review process. The question is to me, whether this is sustainable or the right time to be doing this. Obviously raising money for this is not something that I am not going to do.

I believe there is another way to make a great tool especially in the code review space. Looking back over the last year I see that the majority of the gain in this field has not been from the advances in code understanding but by the increasingly performant agentic abilities of AI models such as Sonnet 3.5, and to some extent GPT 4o. This is not an advancement rate which more money or team members in these previously mentioned companies can alter.

I think there is a space for a more open method which piggybacks off these model advances and those of researchers in the space more than it does to create its own IP. An easily configurable, and flexible platform for code review and understanding, using primarily state-of-the-art foundational models but leaving the door open for locally fine-tuned LLMs. If this package provides the ability to use local models then it should also provide a method to build them easily. That is where I believe that this niche will lead and I want to bet on it.

Last summer I derailed this project trying to build out a SaaS tool. We decided to go with a Serverless specific hosting solution of a GitHub app. This was designed to be able to be easily deployed by us and used by you for some nominal fee. In this new phase I want to take this back to basics. I think we can get a huge amount done as a GitHub action and this will be more approachable to more people. We will not tie ourselves into another hosting provider. I will be building the package code into a container for deployment as an app if necessary but this will be secondary to the main package. This project should be about building the best open project not trying to make a quick buck. 

For the short term I have some UX improvements on my mind for CRGPT. These are in no particular order
- More (or less) than 3 comments on PRs. This solid limit is a hack and should be fixed.
- We should have a better method of judging code which actually looks at the changes as a cohesive mass not as individual files.
- A LGTM comment.
- Line by line for typos and regressions. Not prioritising chat.

Things I will not work on but will be very happy for others to:
- Supporting more foundational LLMs other than GPT4o and Sonnet 3.5
- More platforms such as Gitlab/Azure with a better integration. 

Future upgrades:
- Better algorithms to understand the codebase. If we have to store artifacts we will store it in the codebase itself. 
- Potential to have finetuned models and an ability to make them easily. 
- Better testing and evals. 
- Multiple methods of code understanding and judgement. Users will be able to pick the one they want and contribute new methods. 

If anyone is looking for some open source LLM hobby contributions or maintainer roles please pop me a message. This new phase is going to need more than just me. You will be working for free (at least in the beginning), will need an interest in the space and preferably another job. If that sounds interesting let me know.

I have learnt a huge amount in this last year about code understanding and I am looking forward to build that back into this project. 

Have a great week,
Matt

**Comment by takingstock** [comment]
Hi Matt .. appreciate all the effort you have put into this project so far but most importantly your maturity in accepting the challenges you foresee and your vision. I stumbled upon this space quite late ( i tried like crazy googling projects in this space and couldn't find anything ..tells you a lot about google search results )

i have been hacking away at a project that has a decent overlap with what you have built so far but is more in line with your vision for this project. It would be great if you could go through this  https://takingstock.github.io/codesapper.ai/ .. i have tried to elucidate, as clearly as possible, my views and visions for something comprehensive in this space

if you think there is a good intersection of the venn's i would be more than happy to join you on this journey and contribute ( for free , of course ) .. even i love building stuff ..period

the main reason i would rather jump on to this project is for a very practical reason .. you already have created something that has decent acceptance and velocity and more importantly i am looking for people that are willing to commit to building something useful ( sadly, my network is full of people that are happily settled in life .. no shade at them but thats the main reason i am throwing this hail mary pass ..lol )

**Reply by mattzcarey** [reply]
@takingstock 

Hey thanks for your message. Definitely a bunch of overlap between our interests. Your project seems very similar to some rnd work I was doing [here](https://github.com/mattzcarey/code-review-gpt/pull/341/files#diff-3f65de6dc6a7ead38324d77dcc830e4e107d189027af2e2cd37c5d4afc7f2410).

Very happy if you want to take some leadership on the code mapping generation epic. My current focus is on the easy finetuning and running llms in github actions very efficiently. We could make a rather good team here. Feel free to send me an [email](mailto:matt@stackone.com) and we can chat more. 

Best,
Matt

**Comment by takingstock** [comment]
sent :)

**Comment by wChenonly** [comment]
@mattzcarey  hi, happy to see this project, I would like to try to access more models, but your original implementation is based on @langchain/openai  https://github.com/mattzcarey/code-review-gpt/blob/main/src/common/model/ AIModel.ts#L22. I would like to access other models and I don't think where to add them, can you give me some suggestions?
Thanks for creating this nice and useful project!

**Reply by mattzcarey** [reply]
hey which models would you like. Please open a new discussion thread :)

**Comment by mattzcarey** [comment]
Updates on this. 

We have migrated to a newer tech stack which should be more maintainable. 
We have autoevals and bun test for testing, ai sdk for llm usage. 

I have also removed some specific functionality and cleaned up the main review flow. 

I will be making this flow agentic with a bunch of tools the model can use to get more info about changes. Current tools I am thinking about right now:

- read_file
- suggest_change
- new_file (for proposing tests)
- ask_question 
- shell command?

The aim of the review will be to build a summary of the changes. Along the way it can invoke these tools to review the changes. 
The initial prompt will be scoped down to just the names of the files changed. I'm gonna back that models are getting smarter and we should reduce as much scaffolding as possible.

challenges are always going to be context window. How can we make a good review in one pass. If we need 2 passes how can we collapse the context window in between.

The next version of code review gpt will probably have a generate command 🍄 

also I'm gonna tease the new name `shippie` 🚢 🏴‍☠️
